draft-ietf-tls-ticketrequests.md

```diff
@@ -124,32 +146,71 @@ Clients MAY send this extension in ClientHello. It contains the following struct
 
-Clients can use TicketRequestContents.count to indicate the number of tickets
-they would prefer to receive. Servers SHOULD NOT send more tickets than
-TicketRequestContents.count, as clients will most likely discard any additional
-tickets. Servers SHOULD additionally place a limit on the number of tickets
-they are willing to send, to save resources. Therefore, the number of
-NewSessionTicket messages sent will typically be the minimum of the server's self-imposed
-limit and TicketRequestContents.count.
+struct {
+    select (Handshake.msg_type) {
+        case client_hello: ClientTicketRequest;
+        case encrypted_extensions: ServerTicketRequest;
+    }
+} TicketRequestContents;
+~~~
 
-A server that supports ticket requests MAY echo the "ticket_request" extension in the
-EncryptedExtensions message. If present, it contains a TicketRequestContents structure, where
-TicketRequestContents.count indicates the number of tickets the server expects to send to the client.
+new_session_count
+: The number of tickets desired by the client when the server chooses to
+negotiate a fresh session (full handshake).
+
+resumption_count
+: The number of tickets desired by the client when the server is willing to
+resume using the presented ticket.
+
+expected_count
+: The number of tickets the server expects to send in response to a client
+ticket_request extension.
+
+A client starting a fresh connection SHOULD set new_session_count to the desired
+number of session tickets and resumption_count to 0.
+Once a client's ticket cache is primed, a resumption_count of 1 is a
+good choice that allows the server to replace each ticket with a fresh ticket,
+without over-provisioning the client with excess tickets. However, clients
+which race multiple connections and place a separate ticket in each will
+ultimately end up with just the tickets from a single resumed session.
+In that case, clients SHOULD send a resumption_count equal to the number
+of sessions they are attempting in parallel.
+
+When a client presenting a previously obtained ticket finds that the server
+nevertheless negotiates a fresh session, the client should assume that any
+other tickets associated with the same session as the presented ticket are also
+no longer valid for resumption.  This includes tickets obtained
+during the initial full handshake and all tickets subsequently obtained as
+part of subsequent resumptions.  Requesting more than one ticket in case a
+full handshake is forced by the server helps to keep the session cache primed.
+
+Servers SHOULD NOT send more tickets than requested for the handshake type
+selected by the server (resumption or full handshake). Moreover, servers
+SHOULD place a limit on the number of tickets they are willing to send, whether
+for full handshakes or resumptions, to save resources.  Therefore, the
+number of NewSessionTicket messages sent will typically be the minimum
+of the server's self-imposed limit and the number requested.
```

Comment:
I would add "Servers MAY send additional tickets, up to the same limit, if the tickets that are originally sent are somehow invalidated."